From 7982fb666502f7928c06a8be48046712a51cd7dc Mon Sep 17 00:00:00 2001
From: Sam Shih <sam.shih@mediatek.com>
Date: Wed, 15 Oct 2025 15:05:55 +0800
Subject: [PATCH 11/23] mediatek: add support to verify skip-bad-block BL2
 format

Add support in the Mediatek board code to verify the `skip-bad-block`
format for BL2.

Signed-off-by: Sam Shih <sam.shih@mediatek.com>
CR-Id: WCNCR00310445
Change-Id: If1dc19ec5d78380e766a3ca62d47b978350cb990
---
 board/mediatek/common/bl2_helper.c | 46 ++++++++++++++++++++++++------
 board/mediatek/common/bl2_helper.h |  6 ++--
 2 files changed, 42 insertions(+), 10 deletions(-)

diff --git a/board/mediatek/common/bl2_helper.c b/board/mediatek/common/bl2_helper.c
index eab9f42909c..4e6d2a51b15 100644
--- a/board/mediatek/common/bl2_helper.c
+++ b/board/mediatek/common/bl2_helper.c
@@ -35,6 +35,17 @@ static int bl2_get_compat_list(const void *fdt_blob, struct compat_list *cl)
 	return fdt_read_compat_list(fdt_blob, ret, "bl2_compatible", cl);
 }
 
+static bool bl2_match(const char *compat, struct compat_list cl)
+{
+	int i;
+
+	for (i = 0; i < cl.count; i++)
+		if (!strcmp(compat, cl.compats[i]))
+			return true;
+
+	return false;
+}
+
 static const struct bl2_entry *get_bl2_entry_by_compat(const char *compat)
 {
 	const struct bl2_entry *entry;
@@ -47,17 +58,31 @@ static const struct bl2_entry *get_bl2_entry_by_compat(const char *compat)
 }
 
 static const struct bl2_entry *get_bl2_entry_by_image(const void *bl2,
-						      ulong bl2_size)
+						      ulong bl2_size,
+						      bool skip_badblk_support)
 {
 	const struct bl2_entry *entry;
 	u8 bl2_hdr[BL2_HDR_SIZE];
+	int max_size, offset;
+
+	if (skip_badblk_support) {
+		max_size = BL2_SKIP_BADBLK_COUNT * BL2_SKIP_BADBLK_SIZE;
+		if (max_size >= bl2_size)
+			max_size = bl2_size;
+	} else {
+		max_size = BL2_SKIP_BADBLK_SIZE;
+	}
 
-	memcpy(bl2_hdr, bl2, BL2_HDR_SIZE);
-
-	for (entry = bl2_hdr_entries; entry->compat; entry++) {
-		/* Verify only 8 bytes to prevent BL2_HDR in various version */
-		if (!memcmp(&entry->header, bl2_hdr, 8))
-			return entry;
+	for (offset = 0 ; offset < max_size ; offset += BL2_SKIP_BADBLK_SIZE) {
+		memcpy(bl2_hdr, bl2 + offset, BL2_HDR_SIZE);
+		for (entry = bl2_hdr_entries; entry->compat; entry++) {
+			/*
+			 * Verify only 8 bytes to prevent BL2_HDR in various
+			 * version
+			 */
+			if (!memcmp(&entry->header, bl2_hdr, 8))
+				return entry;
+		}
 	}
 
 	return NULL;
@@ -75,7 +100,12 @@ int bl2_check_image_data(const void *bl2, ulong bl2_size)
 		return 0;
 	}
 
-	entry = get_bl2_entry_by_image(bl2, bl2_size);
+	/* check bl2 have skip bad block support or not */
+	if (bl2_match("spim-nand", cl))
+		entry = get_bl2_entry_by_image(bl2, bl2_size, true);
+	else
+		entry = get_bl2_entry_by_image(bl2, bl2_size, false);
+
 	if (!entry) {
 		cprintln(ERROR, "*** Not a BL2 image ***");
 		goto err;
diff --git a/board/mediatek/common/bl2_helper.h b/board/mediatek/common/bl2_helper.h
index 88d46ddc23f..21175d14fd2 100644
--- a/board/mediatek/common/bl2_helper.h
+++ b/board/mediatek/common/bl2_helper.h
@@ -12,8 +12,10 @@
 #include "upgrade_helper.h"
 
 /* 2MB bl2 max size */
-#define MAX_BL2_SIZE	0x200000
-#define BL2_HDR_SIZE	16
+#define MAX_BL2_SIZE		0x200000
+#define BL2_HDR_SIZE		16
+#define BL2_SKIP_BADBLK_SIZE	0x40000
+#define BL2_SKIP_BADBLK_COUNT	8
 
 /* SF_BOOT */
 #define SPIM_NOR_HDR                                                           \
-- 
2.45.2

